docs/proposals/multiple-dns-provider-support.md

```diff
@@ -0,0 +1,86 @@
+- Add support for multiple providers with a single instance of MGC
+
+## Non Goals
+
+- Support for every DNS provider
+- Support for health checks
+
+## Current Approach
+
+Currently, MGC only supports AWS Route53 as a dns provider. A single instance of a DNSProvider resource is created per MGC instance which is configured with AWS config loaded from the environment. 
+This provider is loaded into all controllers requiring dns access (ManagedZone and DNSRecord reconciliations), allowing a single instance of MGC to operate against a single account on a single dns provider.
+
+## Proposed Solution
+
+MGC has three features it requires of any DNS provider in order to offer full support, DNSRecord management, Zone management and DNS Health checks.  We do not however want to limit to providers that only offer this functionality, so to add support for a provider the minimum that provider should offer is API access to managed DNS records.
+MGC will continue to provide Zone management and DNS Health checks support on a per-provider basis.
+
+Support will be added for AWS(Route53), Google(Google Cloud DNS), Azure and investigation into possible adding CoreDNS (intended for local dev purposes), with the following proposed initial support:
+
+| Provider  | DNS Records | DNS Zones | DNS Health |
+| ------------ | ------------- |  ------------- |---|
+| AWS Route53 | X | X | X |
+| Google Cloud DNS | X | X | - |
+| AzureDNS| X | X | - |
+| CoreDNS| X | - | - |
+
+
+Add DNSProvider as an API for MGC which contains all the required config for that particular provider including the credentials. This can be thought of in a similar way to a cert manager Issuer.
+Update ManagedZone to add a reference to a DNSProvider. This will be a required field on the ManagedZone and a DNSProvider must exist before a ManagedZone can be created.
+Update all controllers load the DNSProvider directly from the ManagedZone during reconciliation loops and remove the single controller wide instance. 
+Add new provider implementations for [google](assets/multiple-dns-provider-support/google/google.md), [azure](assets/multiple-dns-provider-support/azure/azure.md) and coredns.
+    * All providers constructors should accept a single struct containing all required config for that particular provider.
+    * Providers must be configured from credentials passed in the config and not rely on environment variables.
+
+## Testing
+
+{outline any testing considerations. Does this need some form of load/performance test. Are there any considerations when thinking about an e2e test}
+
+## Other Solutions investigated
+
+Investigation was carried out into the suitability of [External DNS] (https://github.com/kubernetes-sigs/external-dns) as the sole means of managing dns resources.
+Unfortunately, while external dns does offer support for basic dns record management with a wide range of providers, there were too many features missing making it unsuitable at this time for integration.
```

Comment:
Was there consideration given to adding the missing features upstream?
What does external-dns roadmap look like and does what we want align?

Reply:
Yes, and we could have forked external-dns and made things work, but decided initially to copy code into our repo and adapt so we could implement faster.

External DNS is quite focused on one part of DNS management, basic record creation, and we likely want to do significantly more in each provider than external dns is likely going to want to add. 

No idea about roadmap of external-dns, but features getting back to main are limited and not much is really added back regularly from what i can gather.